### PR TITLE
Migrate Pausable to component

### LINKED ISSUE #772: Migrate Pausable to component

### PR DESCRIPTION
**Pull request by andrew-fleming**
Fixes #772.

#### PR Checklist

- [x] Tests
- [ ] Tried the feature on a public network
- [ ] Documentation

### CODE REVIEW

File: src/security/pausable.cairo [hunk 1]
```diff
@@ -1,12 +1,17 @@
...
-// OpenZeppelin Contracts for Cairo v0.7.0 (security/pausable.cairo)
-
+/// SPDX-License-Identifier: MIT
+/// OpenZeppelin Contracts for Cairo v0.7.0 (security/pausable.cairo)
+///
+/// # Pausable Component
+///
+/// The Pausable component allows the using contract to implement an
+/// emergency stop mechanism. Only functions that call `assert_paused`
+/// or `assert_not_paused` will be affected by this mechanism.
```
Comment (ericnordelo) [inline src/security/pausable.cairo:8]:
```suggestion
//! SPDX-License-Identifier: MIT
//! OpenZeppelin Contracts for Cairo v0.7.0 (security/pausable.cairo)
//!
//! # Pausable Component
//!
//! The Pausable component allows the using contract to implement an
//! emergency stop mechanism. Only functions that call `assert_paused`
//! or `assert_not_paused` will be affected by this mechanism.
```

File: src/security/pausable.cairo [hunk 2]
```diff
@@ -38,30 +45,48 @@ mod Pausable {
...
+    trait InternalTrait<TContractState> {
+        fn assert_not_paused(self: @TContractState);
+        fn assert_paused(self: @TContractState);
+        fn _pause(ref self: TContractState);
+        fn _unpause(ref self: TContractState);
+    }
+
+    impl InternalImpl<
+        TContractState, +HasComponent<TContractState>
+    > of InternalTrait<ComponentState<TContractState>> {
```
Comment (ericnordelo) [inline src/security/pausable.cairo:67]:
```suggestion
    #[generate_trait]
    impl InternalImpl<
        TContractState, +HasComponent<TContractState>
    > of InternalTrait<TContractState> {
```

File: src/security/pausable.cairo [hunk 3]
```diff
@@ -38,30 +45,48 @@ mod Pausable {
...
+        /// Makes a function callable when the contract is not paused.
```
Comment (ericnordelo) [inline src/security/pausable.cairo:68]:
```suggestion
        /// Makes a function only callable when the contract is not paused.
```

File: src/security/pausable.cairo [hunk 4]
```diff
@@ -38,30 +45,48 @@ mod Pausable {
...
+        fn assert_not_paused(self: @ComponentState<TContractState>) {
             assert(!self.Pausable_paused.read(), Errors::PAUSED);
         }
 
-        fn assert_paused(self: @ContractState) {
+        /// Makes a function callable when the contract is paused.
```
Comment (ericnordelo) [inline src/security/pausable.cairo:73]:
```suggestion
        /// Makes a function only callable when the contract is paused.
```

File: src/security/pausable.cairo [hunk 5]
```diff
@@ -38,30 +45,48 @@ mod Pausable {
...
+        fn assert_paused(self: @ComponentState<TContractState>) {
             assert(self.Pausable_paused.read(), Errors::NOT_PAUSED);
         }
 
-        fn _pause(ref self: ContractState) {
+        /// Triggers a stopped state. The contract must not already be paused.
+        /// Emits a [Paused](Paused) event.
```
Comment (ericnordelo) [inline src/security/pausable.cairo:79]: Is this syntax for events in comments documented?

File: src/tests/mocks/pausable_mock.cairo [hunk 6]
```diff
@@ -0,0 +1,22 @@
+#[starknet::contract]
+mod PausableMock {
+    use openzeppelin::security::pausable::Pausable as pausable_comp;
```
Comment (ericnordelo) [inline src/tests/mocks/pausable_mock.cairo:3]:
```suggestion
    use openzeppelin::security::pausable::Pausable as pausable_component;
```

File: src/tests/mocks/upgrades_v1.cairo [hunk 7]
```diff
@@ -21,10 +21,10 @@ trait UpgradesV1Trait<TState> {
 #[starknet::contract]
 mod UpgradesV1 {
     use array::ArrayTrait;
```
Comment (ericnordelo) [inline src/tests/mocks/upgrades_v1.cairo:23]:
```suggestion
```

File: src/tests/mocks/upgrades_v2.cairo [hunk 8]
```diff
@@ -23,10 +23,10 @@ trait UpgradesV2Trait<TState> {
 #[starknet::contract]
 mod UpgradesV2 {
     use array::ArrayTrait;
```
Comment (ericnordelo) [inline src/tests/mocks/upgrades_v2.cairo:25]:
```suggestion
```

File: src/security/pausable.cairo (hunk 5 above)
Comment (andrew-fleming) [reply src/security/pausable.cairo:79]: No, good catch. Will update

**Review by ericnordelo — APPROVED** [review]
LGTM!

**Pull request merged**